File: src/utils/placeholders.ts
/**
 * 替换占位符，同时支持两种格式：
 * - {{user}} / {{char}} (SillyTavern 标准)
 * - <user> / <char> (部分类脑社区卡片使用)
 */
export function replacePlaceholders(
  text: string,
  charName: string,
  userName: string
): string {
  return text
    .replace(/\{\{char\}\}/gi, charName)
    .replace(/\{\{user\}\}/gi, userName)
    .replace(/<char>/gi, charName)
    .replace(/<user>/gi, userName);
}

File: fake/strange/backend/app/prompt/placeholders.py
"""占位符替换，同时支持 {{user}}/{{char}} 与 <user>/<char>。"""
import re

_PATTERNS = [
    (re.compile(r"\{\{char\}\}", re.IGNORECASE), "char"),
    (re.compile(r"\{\{user\}\}", re.IGNORECASE), "user"),
    (re.compile(r"<char>", re.IGNORECASE), "char"),
    (re.compile(r"<user>", re.IGNORECASE), "user"),
]


def replace_placeholders(text: str, char_name: str, user_name: str) -> str:
    result = text
    for pattern, kind in _PATTERNS:
        replacement = char_name if kind == "char" else user_name
        result = pattern.sub(lambda _m: replacement, result)
    return result
